refactor: Revamp CI workflows

bthomee opened the pull request:
## High Level Overview of Change

This PR refactors the CI workflows to leverage the new CI Docker images for Debian, Red Hat, and Ubuntu. All workflows and actions have been rewritten as follows.

**On each PR commit**

The following jobs run:
* Check clang-format: same functionality as current CI.
* Check levelization: same functionality as current CI, but scripts have been moved into the `.github` directory.
* Build and Test:
  * Linux: Runs Debian, Red Hat, and Ubuntu in various configurations. Some configurations are used to test specific operations, such as setting the reference fee to a different value, and codecov.
  * MacOS.
  * Windows.
  * Before building and testing, any needed dependencies are pulled from the Conan remote, and any missing ones are installed. In PR commits these dependencies are _not_ uploaded.
* Notify Clio: similar functionality as current CI, except that we now use the `needs:` keyword to wait on the Linux, MacOS, and Windows jobs instead of using a third-party action to poll for the completion of only the Linux jobs.

** On trigger**

The following jobs run for the following events:
* Schedule: Each work day the dependencies are fully rebuilt and any missing ones are uploaded.
  * This is just as a sanity check, and is similar to how Clio does this.
  * In normal circumstances the rebuilt dependencies should be exactly the same as what is available in the Conan remote. However, if a Conan recipe received a patch in our Conan Center fork, it is possible for the rebuilt dependency to be different. Also, if the Conan remote has received maintenance and some packages were deleted, this ensures they are restored.
* Workflow dispatch: This is triggered manually using the GitHub UI or API.
  * It supports two "force" flags, to rebuild all dependencies and to upload (=overwrite) the dependencies.
* Push: When a PR is merged into the develop, release, or master branch the Build and Test workflows are run, as described in the section above, but this time any missing dependencies are uploaded.
  * This ensures that when a dependency is updated it will be added to the Conan remote upon merge of an approved PR into the develop branch.
  * The uploading is also enabled when merging into the release and master branch, since in rare cases a hotfix may result in a dependency change.

### Context of Change

The CI workflows needed some TLC after the Conan 2 update and the new availability of our Conan Center Index fork.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

## Future Tasks
It can be considered to add extra labels that, when applied to a PR, will trigger a specific run using one or more specific configurations. For instance, if a configuration using Clang 19 fails to build we may only notice this during merge; being able to trigger this during a PR will facilitate debugging.

Bronek commented on .github/scripts/strategy-matrix/generate.py:
```diff
@@ -0,0 +1,141 @@
+            if os['distro_name'] == 'windows' and not (build_type == 'Release' and '-Dunity=ON' in cmake_args and architecture['platform'] == 'windows/amd64'):
+                continue
+
+
```
I would add `else:` section here, so we do not run both unity and non-unity builds for every configuration; this means for most cases (macos, windows, selection of Linux configs) skip non-unity builds, and for the opposite selection of Linux configs, skip unity builds.

Bronek reviewed (commented):
Looking good, some minor comments but I have not finished yet.

bthomee commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+        conan remote login ${{ inputs.conan_remote_name }} "${{ inputs.conan_remote_username }}" --password "${{ inputs.conan_remote_password }}"
+        echo 'Uploading dependencies.'
+        conan list --graph=build.json --graph-binaries=build --format=json > pkgs.json
+        conan upload --list=pkgs.json --confirm --check ${{ inputs.force_upload && '--force' || '' }} --remote=${{ inputs.conan_remote_name }}
```
It's an issue on the self-hosted Mac (and soon Windows) runners, where it will upload everything - even if modified by a completely unrelated PR. As such, to err on the side of caution I prefer to keep it this way.

mathbunnyru replied:
https://github.com/XRPLF/clio/blob/develop/.github/actions/prepare_runner/action.yml

bthomee replied:
I've made this change with `CONAN_HOME` for both MacOS and Windows.

bthomee replied:
After setting `CONAN_HOME`, I changed back to do `conan upload '*'` and removed the use and upload of `build.json` and `pkgs.json`

Bronek commented on .github/scripts/strategy-matrix/linux.json:
```diff
@@ -0,0 +1,146 @@
+      "distro_version": "bookworm",
+      "compiler_name": "gcc",
+      "compiler_version": "14"
+    },
```
can we add here this ?

```
    {
      "distro_name": "debian",
      "distro_version": "bookworm",
      "compiler_name": "gcc",
      "compiler_version": "15"
    },
```

bthomee replied:
Done

Bronek commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+      run: |
+        echo 'Installing configuration and profile.'
+        if [ '${{ inputs.os }}' == 'Linux' ]; then
+          echo '${{ inputs.conan_global_conf }}' >> $(conan config home)/global.conf
```
is there some `global.conf` on MacOS or Windows that we want to overwrite (I know there is one on Linux that we want to keep) ?

if not, then perhaps we can make this `>>` unconditional ?

Bronek commented on .github/scripts/strategy-matrix/generate.py:
```diff
@@ -0,0 +1,141 @@
+            if os['distro_name'] == 'debian':
+                skip = True
+                if os['distro_version'] == 'bookworm':
+                    if f'{os['compiler_name']}-{os['compiler_version']}' == 'gcc-12' and build_type == 'Debug' and '-Dunity=ON' in cmake_args and architecture['platform'] == 'linux/amd64':
```
Another reason why this should be non-unity : coverage reporting from a unity build would be nonsensical.

Bronek commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+      run: |
+        echo 'Configuring CMake.'
+        cmake \
+          ${{ inputs.cmake_generator && format('-G "{0}"', inputs.cmake_generator) || '' }} \
```
Do we need this long expression here ? It seems like we always set `cmake_generator`

Bronek commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+          echo 'The binary is statically linked.'
+        else
+          echo 'The binary is dynamically linked.'
+        exit 1
```
misleading indentation.

bthomee replied:
Thanks, will fix.

Bronek commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+        plugins: noop
+        token: ${{ inputs.codecov_token }}
+        verbose: true
+    - name: Test the binary
```
we can skip this step if target was `coverage`, because running unit tests is part of `coverage` target (there's no coverage report without unit tests being run).

Bronek replied:
it only runs the "old style" `./rippled -u` unit tests.

bthomee replied:
How much effort will it be to have it run both? Eventually the vast majority of tests will be out of the rippled binary and only run as part of ctest.

Bronek commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+        echo 'Verifying presence of instrumentation.'
+        ./rippled --version | grep libvoidstar
+    - name: Upload coverage report
+      if: ${{ contains(inputs.cmake_args, '-Dcoverage=ON') && inputs.codecov_token }}
```
I would also swap the order of steps - "Test the binary" first (if target is _not_ `coverage`) first, then "Upload coverage report" (if target is `coverage`)

bthomee replied:
Done

Bronek commented on .github/scripts/strategy-matrix/generate.py:
```diff
@@ -0,0 +1,141 @@
+        # for Windows, but it can get overridden for certain configurations.
+        cmake_target = 'install' if os["distro_name"] == 'windows' else 'all'
+
+        # Only run a subset of configurations in PRs.
+        if pr:
```
I disagree; it would be simply too many - we are talking multidimensional product of distro/compiler/build type/unity on Linux alone, where we support 9 different compilers (a subset of it for some distros) and 3 distros. On top of that we have MacOS and Windows. 

Just a product of build type/unity build is 4 (i.e. Release/unity , Release/non-unity , Debug/unity , Debug/non-unity) then multiply that by the above and we are getting too many builds.

mathbunnyru replied:
For dependencies related changes we should build every configuration (if we should build dependencies at all).
Number of such changes would be relatively small (after we finish working on Conan)

bthomee commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+        conan remote login ${{ inputs.conan_remote_name }} "${{ inputs.conan_remote_username }}" --password "${{ inputs.conan_remote_password }}"
+        echo 'Uploading dependencies.'
+        conan list --graph=build.json --graph-binaries=build --format=json > pkgs.json
+        conan upload --list=pkgs.json --confirm --check ${{ inputs.force_upload && '--force' || '' }} --remote=${{ inputs.conan_remote_name }}
```
Done - I had to ensure each configuration has a unique name based on the matrix, so I modified the strategy matrix generation script accordingly. I'm also uploading the `build.json` for good measure. I set a 7-day retention period, thinking that will be sufficient.

Bronek replied:
I would prefer much longer than 7 days, for troubleshooting purposes "huh we broke something ... when did it happen ?"

bthomee replied:
Done, changed it to 30 days.

bthomee commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+      working-directory: ${{ inputs.build_dir }}
+      run: |
+        echo 'Building binary.'
+        cmake --build . \
```
Done.

bthomee commented on .github/actions/build-test/action.yml:
```diff
@@ -0,0 +1,101 @@
+        ./rippled --version | grep libvoidstar
+    - name: Upload coverage report
+      if: ${{ contains(inputs.cmake_args, '-Dcoverage=ON') && inputs.codecov_token }}
+      uses: codecov/codecov-action@18283e04ce6e62d37312384ff67231eb8fd56d24
```
On the right side... of each `with:` option you mean?

bthomee commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+      shell: bash
+      run: |
+        echo 'Installing configuration and profile.'
+        if [ '${{ inputs.os }}' == 'Linux' ]; then
+          echo '${{ inputs.conan_global_conf }}' >> $(conan config home)/global.conf
+        else
+          echo '${{ inputs.conan_global_conf }}' > $(conan config home)/global.conf
+        fi
```
Regarding cleaning the Conan directory, I'm concerned about parallel runs upon merge, where we'll have Debug+Release x Unity+Non-Unity x amd64+arm64...

mathbunnyru replied:
What exactly is your concern?

bthomee replied:
One workflow is running, it clears `~/.conan2` and starts building dependencies there for amd64. Then another workflow starts, also clears the shared `~/.conan2`, and then the first workflow doesn't have the dependencies it built anymore.

mathbunnyru replied:
> The discrepancy is because each build image will has its own `global.conf`, with:
> 
> ```
> tools.build:compiler_executables={"c": "${CC}", "cpp": "${CXX}"}
> user.package:distro="debian-${DEBIAN_VERSION}"
> tools.info.package_id:confs=["user.package:distro"]
> ```
> 
> ... and we really want to keep these. EDIT: there could be also `distro="rhel-..."` or `distro="ubuntu-..."`.

Why not move these to a profile instead of having it in `global.conf`?

bthomee replied:
Can you elaborate? I'm pretty sure I see more jobs running in parallel than we have machines.

Bronek replied:
I think there is distinction between "running within a container" (like all our Linux jobs do) and "running on bare metal" (which is MacOS and Windows, since there's no containers for these). When running in a container, I see no reason why multiple jobs wouldn't run in parallel. When running on bare metal, I imagine github runner might give exclusive access to the machine, to avoid the complexities of providing isolation between jobs in some other way.

mathbunnyru commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+        conan remote login ${{ inputs.conan_remote_name }} "${{ inputs.conan_remote_username }}" --password "${{ inputs.conan_remote_password }}"
+        echo 'Uploading dependencies.'
+        conan list --graph=build.json --graph-binaries=build --format=json > pkgs.json
+        conan upload --list=pkgs.json --confirm --check ${{ inputs.force_upload && '--force' || '' }} --remote=${{ inputs.conan_remote_name }}
```
> have a separate workflow for PR commits where uploading is disabled, and one for PR merges, schedules, and manual triggers where uploading is enabled

If the first one builds rippled code, and second doesn't (and only builds dependencies) - that's exactly what we have in Clio.
And you will be able to move `conan upload` to this workflow, as I suggested

bthomee commented:
Latest updates:
* I merged the `build-linux`, `build-macos`, and `build-windows` jobs and inserted OS-specific modifications as needed.
* I tried moving the `CONAN_REMOTE_NAME` and `CONAN_REMOTE_URL` into the `build-test` action and the `notify-clio` workflow. However, in the action the environment variables were empty, so the jobs failed, so I restored it the way it was.
  * There is generally an issue whereby `inputs`, `secrets`, and `env` are available only in certain circumstances, and the documentation is quite poor and the reasoning equally so.
* I use `type: choice` wherever possible.
  *  Similar to above, this type of input is only available in certain places. It is in actions, but it isn't in reusable workflows.

mathbunnyru commented on .github/workflows/on-trigger.yml:
```diff
@@ -0,0 +1,147 @@
+# This workflow runs all workflows to build the dependencies required for the
```
This workflow also highly duplicates the previous one and we should merge them in a single workflow

Bronek replied:
FWIW I do agree with @bthomee - these two workflows have different responsibilities (if you look outside of build/test cycle) hence should be separate.

Bronek commented on .github/actions/build-deps/action.yml:
```diff
@@ -0,0 +1,108 @@
+    required: true
+    type: string
+  conan_global_conf:
+    description: 'The contents of the global Conan configuration file.'
```
FWIW we know that this string is really small (under 1KB) and it only becomes a file _inside_ this action.

mathbunnyru commented on .prettierignore:
```diff
@@ -0,0 +1,2 @@
+external
+.*
```
Why do you need this?
It looks like `external` is actually now code written by us, so it should be prettified.

mathbunnyru replied:
Ok, then please revert changes from https://github.com/XRPLF/rippled/pull/5657 done to this folder
And I think it's better to be done in a separate PR (including changes to this file as well)

Bronek replied:
Duh I totally missed that PR :-|

Bronek replied:
@bthomee I think @mathbunnyru is correct, we should revert changes from that other commit under `external`

bthomee commented on conan/windows.conf:
```diff
@@ -0,0 +1,9 @@
+  # Global configuration for Conan. This is used to set the number of parallel
```
We currently have different settings for Windows. For simplicity I kept them the same here and planned to tune them after.

mathbunnyru replied:
1. I don't think they will be often tuned
2. If they are, it can still be done in a single file: https://docs.conan.io/2/reference/config_files/global_conf.html#configuration-file-template

Bronek commented on .github/workflows/build-test.yml:
```diff
@@ -0,0 +1,186 @@
+      os:
+        description: 'The operating system to use for the build (linux, macos, or windows).'
+        required: true
+        type: string
```
let's make it `choice`, like in `build-test/action.yml`

Bronek replied:
Thanks for trying 😢

Bronek commented on .github/workflows/on-pr.yml:
```diff
@@ -0,0 +1,110 @@
+      - '.github/workflows/check-levelization.yml'
+      - '.github/workflows/notify-clio.yml'
+      - '.github/workflows/on-pr.yml'
+      # Keep the list of paths below in sync with those in `on-trigger.yml`.
```
would it be simpler to use the reverse i.e. prefix `!` for exclusion of certain files ? https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#onpushpull_requestpull_request_targetpathspaths-ignore

Bronek commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,53 @@
+      - '**/*.md'
+      - 'docs/**'
+      - 'include/**'
+      - 'src/**'
```
I think we also need `workflow_dispatch:`

bthomee replied:
I'll be happy to do so, but in which circumstances would we need to republish the docs? Upon PR merge into `develop` the job will run, and will publish them.

Bronek commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,53 @@
+          cmake -Donly_docs=ON ..
+          cmake --build . --target docs --parallel $(nproc)
+      - name: Publish documentation
+        if: ${{ github.ref_type == 'branch' && github.ref_name == github.event.repository.default_branch }}
```
we should also allow this to work on `workflow_dispatch:`

bthomee replied:
(See also previous comment.) If adding `workflow_dispatch`, then it is necessary to select the branch it runs on. In those cases, `ref_type` will be `branch` and `ref_name` will be `develop` - exactly what we want.

If we just allow the publish to run on any `workflow_dispatch` event, someone could select a different branch and have the docs there get published instead - which would presumably not be what we want.

QQ: Where exact are the docs published to? (We also should clean up the `docs/` directory in the repo, it has a lot of stuff that's outdated.)

Bronek replied:
> If we just allow the publish to run on any workflow_dispatch event, someone could select a different branch and have the docs there get published instead - which would presumably not be what we want.

excellent point. We _could_ enforce that this fails unless the selected branch is one of the 3 major `master` `release` `develop`

Bronek replied:
actually you are right, let's not do that. Worst case, we will merge "dummy" change into `develop` if we need to trigger this workflow.

bthomee commented on .github/workflows/on-trigger.yml:
```diff
@@ -0,0 +1,137 @@
+  # all dependencies can be built successfully. Only the dependencies that
+  # are actually missing from the remote will be uploaded.
+  schedule:
+    - cron: '32 7 * * 1-5'
```
Will do in the next commit.

bthomee commented on .github/workflows/notify-clio.yml:
```diff
@@ -0,0 +1,101 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.clio_notify_token }}
+        run: |
+          echo "Fetching PR number associated with current commit '${{ github.sha }}'."
```
I ran some more experiments, but the PR number is not included in that variable. See https://github.com/XRPLF/rippled/pull/5692. I even printed out the entire object but the PR number was not included on the `push` event. The current solution isn't so bad in the end, luckily.

bthomee commented:
I'll merge this after 2.6.0 has been released to not cause any disruptions during the release process.

bthomee commented:
I made some small tweaks to further improve the workflows, and I think they're now in a good place.
* The `on-pr.yml` workflow now uses `on: pull_request`.
  * That event has proven to be quite fragile, but it's now working.
  * I removed the `on: push` as this caused duplicate workflow invocations.
  * I removed the `branches` from the `on: pull_request` as we still want the workflows to run for other base branches too, e.g. the smart escrows branch or when you're working on a chain of dependent PRs.
  * I improved the logic for when the workflow should be triggered. In short, you no longer need to close and reopen a PR to get it to trigger, but can simply remove and re-add the `DraftCIRun` label as the logic now also triggers on the `labeled` and `unlabeled` events - but intentionally only for the `DraftCIRun` label and not for other labels.
* Simplified the strategy matrix script to take an `--all` flag instead of having to work out itself based on the branch whether to generate all configurations or a subset.

As for follow-ups:
* Simplify the strategy matrix generation to remove the .json files and handle it wholly within the .py file using the cross-product of all possible combinations and then filtering them down to the actually viable ones.
* Consider adding support for triggering configurations to be built for PR commits that are otherwise excluded.
  * For instance, we know that the RHEL builds currently fail on ARM64, so it would be nice to be able to add a label or some other mechanism to explicitly trigger them on PR commits to test whether a proposed fix works.
